src/plan_add_hashagg.c

```diff
@@ -0,0 +1,575 @@
+#include <optimizer/paths.h>
+#include <optimizer/tlist.h>
+#include <catalog/namespace.h>
+#include <catalog/pg_type.h>
+#include <utils/guc.h>
+#include <miscadmin.h>
+#include <utils/timestamp.h>
+#include <utils/lsyscache.h>
+#include <utils/selfuncs.h>
+
+#include "compat-msvc-enter.h"
+#include <optimizer/cost.h>
+#include "compat-msvc-exit.h"
+
+#include "planner_import.h"
+#include "plan_add_hashagg.h"
+#include "extension.h"
+#include "utils.h"
+#include "guc.h"
+#include "dimension.h"
+#include "chunk_dispatch_plan.h"
+#include "planner_utils.h"
+#include "hypertable_insert.h"
+#include "constraint_aware_append.h"
+#include "compat.h"
+
+#define INVALID_ESTIMATE -1
+#define IS_VALID_ESTIMATE(est) (est >= 0)
+#define MAX_FUNCTION_ARGS 10
+
+static double custom_group_estimate_time_bucket(PlannerInfo *root, FuncExpr *expr, double path_rows);
+static double custom_group_estimate_date_trunc(PlannerInfo *root, FuncExpr *expr, double path_rows);
+
+typedef struct
+{
+	Oid			function_oid;
+	char	   *function_name;
+	int			nargs;
+	double		(*custom_group_estimate_func) (PlannerInfo *root, FuncExpr *expr, double path_rows);
+	Oid			arg_types[MAX_FUNCTION_ARGS];
+} CustomEstimateForFunctionInfo;
+
+/* Definitions of functions with a custom group estimate.
+   These will also be entered into the custom_estimate_func_hash hash table. */
+static CustomEstimateForFunctionInfo custom_estimate_func_info[] =
+{
+	{
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, TIMESTAMPOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket
+	},
+	{
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, TIMESTAMPTZOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket,
+	},
+	{
+		.function_name = "time_bucket",
+		.nargs = 2,
+		.arg_types = {INTERVALOID, DATEOID},
+		.custom_group_estimate_func = custom_group_estimate_time_bucket,
+	},
+	{
+		.function_name = "date_trunc",
+		.nargs = 2,
+		.arg_types = {TEXTOID, TIMESTAMPOID},
+		.custom_group_estimate_func = custom_group_estimate_date_trunc,
+	},
+	{
+		.nargs = 2,
+		.custom_group_estimate_func = custom_group_estimate_date_trunc,
+		.arg_types = {TEXTOID, TIMESTAMPTZOID},
+		.function_name = "date_trunc",
+	},
+};
+#define _MAX_HASHAGG_FUNCTIONS (sizeof(custom_estimate_func_info)/sizeof(custom_estimate_func_info[0]))
+
+HTAB	   *custom_estimate_func_hash = NULL;
+
+static bool
+function_types_equal(Oid left[], Oid right[], int nargs)
+{
+	int			arg_index;
+
+	for (arg_index = 0; arg_index < nargs; arg_index++)
+	{
+		if (left[arg_index] != right[arg_index])
+			return false;
+	}
+	return true;
+}
+
+static void
+initialize_custom_estimate_func_info()
+{
+	int			i = 0;
+
+	HASHCTL		hashctl = {
+		.keysize = sizeof(Oid),
+		.entrysize = sizeof(CustomEstimateForFunctionInfo)
+	};
+
+	custom_estimate_func_hash = hash_create("custom_estimate_func_hash", _MAX_HASHAGG_FUNCTIONS, &hashctl, HASH_ELEM | HASH_BLOBS);
```

Comment:
Any reason to not use our standard "cache" implementation here and stick to "common" code?

Reply:
Because we don't need pinning or invalidation and this code is easy. Do you think we should use the cache instead?

Reply:
Looking at `hypertable_cache_create`, the code there seems more complex than the usage here, so I don't see a reason to switch to the Cache implementation